keps/sig-scheduling/20190311-resource_bin_packing_priority_function.md

```diff
@@ -0,0 +1,167 @@
+---
+title: Resource Bin Packing Priority Function
+  - "@sudeshsh"
+owning-sig: sig-scheduling
+participating-sigs:
+  - sig-scheduling
+reviewers:
+  - "@k82cn"
+  - "@Huang-Wei"
+approvers:
+  - "@k82cn"
+creation-date: 2019-03-11
+last-updated: 2019-03-11
+status: provisional
+---
+
+# Resource Bin Packing Priority Function
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+- [Design Details](#design-details)
+  - [Examples](#examples)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+## Summary
+
+Resource Bin Packing Priority Function allows users to use the best fit polices during scheduling. It allows users to apply bin packing on core resources like CPU, Memory as well as extended resources like accelerators.
+
+## Motivation
+
+While running Machine Learning workloads on kubernetes which use accelerator devices the default scheduler spreads pods across nodes resulting in fragmentation of extended resources. This fragmentation prevents scheduling pods with larger device resource requirements, and they remain in pending state.
+
+### Goals
+
+- Schedule Jobs using BestFit Policy using Resource Bin Packing Priority Function
+- Reduce Fragmentation of scarce resources on the Cluster
+
+### Non-Goals
+
+-
+
+## Proposal
+
+The plan is to add resource_bin_packing  as optional priority function. Add another argument resources of `type map[v1.ResourceName]int64{}` .This would allow users who want to bin pack a resource to use the function by setting the argument resources which would require them to specify weights for bin packing. For example
```

Comment:
We already have a priority function for bin packing, called "MostRequestedPriority". We also have another priority function called "RequestedToCapacityRatioResourceAllocationPriority" which is a configurable resource to capacity priority function. It can be configured to act as "most requested", "least requested", or anything in between. Both of the existing priority functions support only CPU and Memory. It would make more sense if we could extend one of these two to support extended resources instead of adding another priority function that functions very similarly to the existing ones. Of course, we need to ensure that our changes won't break backward compatibility.

Reply:
Yes, I will update the KEP to indicate the changes that need to be done to "RequestedToCapacityRatioResourceAllocationPriority" and "ResourceAllocationPriority "which would broaden the scope of these functions to allow extended resources and weights to each resource being requested.
Will make sure that this still does not break the  "MostRequestedPriority"

Reply:
@sudeshsh Thanks for updating the KEP, but a lot of places, including your example config, are still about adding a new priority function. I think the KEP should be more explicit about the changes needed to be made to the RequestedToCapacityRatioResourceAllocationPriority to make it usable for scarce resource bin packing.

Reply:
@bsalamat As per my understanding, `MostRequestedPrioirty` and `LeastRequestedPriority` are internally calling `RequestedToCapacityRatioResourceAllocationPriority`, But to set these priorities we still need to call them as Individual priority function as pointed here https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/api/compatibility/compatibility_test.go#L281.  As mentioned I will integrate most of the code for BinPacking inside `RequestedToCapacityRatioResourceAllocationPriority` but the config file to set this priority would remain the same as mentioned in the example. I will add more details about changes to be done to `RequestedToCapacityRatioResourceAllocationPriority`

Reply:
No, MostRequestedPrioirty and LeastRequestedPriority do not call RequestedToCapacityRatioResourceAllocationPriority. Those two predate `RequestedToCapacityRatioResourceAllocationPriority`. `RequestedToCapacityRatioResourceAllocationPriority` is not enabled by default. `LeastRequestedPriority` is the only one enabled by default.
Integrating your proposed changes into `RequestedToCapacityRatioResourceAllocationPriority` SGTM.

Reply:
@bsalamat I updated the KEP as per the comments.